openedge-plugin/src/main/java/org/sonar/plugins/openedge/sensor/OpenEdgeProparseSensor.java

```diff
@@ -57,156 +62,227 @@
 import org.sonar.plugins.openedge.api.org.prorefactor.treeparser.ParseUnit;
 import org.sonar.plugins.openedge.api.org.prorefactor.treeparser.SymbolScope;
 import org.sonar.plugins.openedge.foundation.CPDCallback;
+import org.sonar.plugins.openedge.foundation.IIdProvider;
 import org.sonar.plugins.openedge.foundation.OpenEdgeComponents;
 import org.sonar.plugins.openedge.foundation.OpenEdgeMetrics;
+import org.sonar.plugins.openedge.foundation.OpenEdgeProjectHelper;
 import org.sonar.plugins.openedge.foundation.OpenEdgeRulesDefinition;
 import org.sonar.plugins.openedge.foundation.OpenEdgeSettings;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
 
 public class OpenEdgeProparseSensor implements Sensor {
   private static final Logger LOG = Loggers.get(OpenEdgeProparseSensor.class);
 
   // IoC
   private final OpenEdgeSettings settings;
   private final OpenEdgeComponents components;
+  private final IIdProvider idProvider;
+
+  // Internal use
+  private final DocumentBuilderFactory dbFactory;
+  private final DocumentBuilder dBuilder;
 
-  public OpenEdgeProparseSensor(OpenEdgeSettings settings, OpenEdgeComponents components) {
+  // File statistics
+  private int numFiles;
+  private int numXREF;
+  private int numFailures;
+  // Timing statistics
+  private Map<String, Long> ruleTime = new HashMap<>();
+  private long parseTime = 0L;
+  private long xmlParseTime = 0L;
+  private long maxParseTime = 0L;
+  // Proparse debug
+  List<String> debugFiles = new ArrayList<>();
+
+  public OpenEdgeProparseSensor(OpenEdgeSettings settings, OpenEdgeComponents components, IIdProvider idProvider) {
     this.settings = settings;
     this.components = components;
+    this.idProvider = idProvider;
+
+    this.dbFactory = DocumentBuilderFactory.newInstance();
+    try {
+      this.dBuilder = dbFactory.newDocumentBuilder();
+    } catch (ParserConfigurationException caught) {
+      throw new RuntimeException(caught);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS00112)